File: ProblemSolving/ProblemSolving/PG67257_수식최대화.cpp
#include <iostream>
#include <string>
#include <vector>
#include <stack>
using namespace std;

// 0, 1, 2 순열 구하기

char arr[3] = { '+', '-', '*' };
string str;

void swap(int index1, int index2) {
    int tmp = arr[index1];
    arr[index1] = arr[index2];
    arr[index2] = tmp;
}

void dfs(int depth) {
    if (depth == 3) {
        
        // 일단 문자열 하나씩 접근해서 숫자면 이전 숫자에 덧붙이기
        // 연산자가 나오면, 지금 우선순위인 연산자인지 확인
        // 맞으면
            // flag 활성화
            // 다음 숫자를 만들어서 이전 숫자와 연산 후 stack에 넣기, flag 비활성화
        // 다르면 
            // 지금까지 만든 숫자 stack에 넣기
            // 연산자 stack에 넣기

        stack<string> s;
        
        string preNum = "";
        for (int i = 0; i < str.size(); i++) {
            
        }



        return;
    }

    for (int i = depth; i < 3; i++) {
        swap(depth, i);
        dfs(depth + 1);
        swap(depth, i);
    }
}

long long solution(string expression) {
    long long answer = 0;

    // 연산자 우선순위 정하기
    
    str = expression;
    dfs(0);
    

    return answer;
}

int main() {

    cout << solution("100-200*300-500+20");

    return 0;
}

File: ProblemSolving_Java/ProblemSolving_Java/src/programmers/level2/PG67257_수식최대화.java
package programmers.level2;

import java.util.ArrayList;
import java.util.List;

public class PG67257_수식최대화 {

    public static void main(String[] args) {

        String expression = "100-200*300-500+20";
        Solution solution = new Solution();
        long solution1 = solution.solution(expression);
        System.out.println("solution1 = " + solution1);
    }

    static class Solution {

        String[] pick = {"+", "-", "*"};
        List<String> parsedList = new ArrayList<>();
        Long maxVal = -1L;

        public long solution(String expression) {
            long answer = 0;

            // 연산자, 피연산자를 배열에 담고
            String pre = "";
            for(int i=0; i<expression.length(); i++) {
                char ch = expression.charAt(i);
                if(ch == '+'
                        || ch == '-'
                        || ch == '*') {
                    parsedList.add(pre);
                    parsedList.add(ch+"");
                    pre="";
                }
                else
                    pre += ch;
            }
            parsedList.add(pre);

            // 연산자 우선순위를 순열로 구해서 계산
            SetPriority(0);

            return maxVal;
        }

        private void SetPriority(int depth) {

            if(depth == 3) {

                List<String> opList = new ArrayList<>(parsedList);
                for(int k=0; k<3; k++) {
                    List<String> tmp = new ArrayList<>();

                    int idx = 0;
                    while(idx < opList.size()) {
                        if(opList.get(idx).equals(pick[k])) {
                            tmp.set(tmp.size()-1, Long.toString(calculate(tmp.get(tmp.size()-1), pick[k], opList.get(idx+1))));
                            idx+=2;
                        }
                        else {
                            tmp.add(opList.get(idx));
                            idx++;
                        }
                    }

                    opList = tmp;
                }

                Long value = Math.abs(Long.parseLong(opList.get(0)));

                if(maxVal < value)
                    maxVal = value;

                return;
            }

            for(int i=depth; i<3; i++) {
                swap(i, depth);
                SetPriority(depth+1);
                swap(i, depth);
            }
        }

        private void swap(int idx1, int idx2) {
            String tmp = pick[idx1];
            pick[idx1] = pick[idx2];
            pick[idx2] = tmp;
        }

        private Long calculate(String op1, String operator, String op2) {

            Long n1 = Long.parseLong(op1);
            Long n2 = Long.parseLong(op2);

            if(operator.equals("+"))
                return n1+n2;
            else if(operator.equals("-"))
                return n1-n2;
            else
                return n1*n2;
        }
    }

}
